### DISCUSSION TITLE
cipher 的翻譯

### DISCUSSION THREAD

**Discussion by mattwang44** [python/python-docs-zh-tw#771] [Add Term]
### 初步確認

- [X] 我已經確認過 [術語列表](https://github.com/python/python-docs-zh-tw/wiki/%E8%A1%93%E8%AA%9E%E5%88%97%E8%A1%A8) 中沒有相關術語。
- [X] 我已經確認過 [Python 官方術語列表](https://docs.python.org/zh-tw/3/glossary.html) 中沒有相關術語。

### 術語原文

cipher

### 術語翻譯

TBD

### 說明

特指 cipher 作為加密用的 secret 時應該被翻譯成什麼，「加密方法」相關的意思則不在這次討論範圍內。

- 出處：多處文件出現，如 library/ssl.po
- 說明：現有繁中文章大多翻譯成「密碼」，但容易讓讀者和 `secret` 一詞混淆。覺得使用「暗碼」一詞也許會相對好辨認，但目前並不是很通用。

### 參考資料

- [教育國語辭典的「暗碼」頁面](https://dict.concised.moe.edu.tw/dictView.jsp?ID=39794&la=0&powerMode=0)
- 有一本書名為 "The Cipher" 的書籍是被翻譯成「暗碼」（[博客來頁面](https://www.books.com.tw/products/0010967230)）

**Comment by rockleona** [comment]
之前在資安公司待過，有用到「密語」一詞，雖然好像還是不大通用，晚點補上參考資料

**Reply by rockleona** [reply]
https://hitcon.org/2018/CMT/slide-files/d1_s2_r4.pdf

**Comment by ken71301** [comment]
3/30：目前觀察翻譯原文，secert 應翻譯為 `機密`，cipher應翻譯為 `密碼`

**Comment by SkyLull** [comment]
是否能夠請 @mattwang44 再解釋得清楚一點，`cipher 作為加密用的 secret` 具體而言出現在哪裡？

因為 `cipher` 本身主要就是加密方法的意思 [[牛津辭典](https://www.oxfordlearnersdictionaries.com/definition/english/cipher)] [[劍橋辭典](https://dictionary.cambridge.org/dictionary/english/cipher)]
其中[[牛津辭典](https://www.oxfordlearnersdictionaries.com/definition/english/cipher)]的解釋倒是提出另一種語意 `可數名詞。一個不重要的人或事物。`

如果被作為 `secret` 使用，是不是有該文章語境下的特殊情況？是否能直接將其視為 `secret` 並翻譯為`機密`？

您也提到本串 issue 並不討論 `cipher` 作為「加密方法」的這個一般狀況，而是您提出的特例，但這個特例的結論卻被加入到術語表當中作為通則了。

這個討論的結果是否**應在術語表中加上額外註解**，或**不應該出現在術語表中**？

---

此外，在您提出的 `ssl.po` 中，其實已經出現多次將 `cipher` 譯為「密碼」而非「加密方法」的狀況。雖然觀察其撰寫時間點並非由術語表造成的誤譯，但仍令人擔心是否會影響未來翻譯其他使用該字的文件。

**Reply by mattwang44** [reply]
Hi @SkyLull,

這詞討論就是我在 review `ssl.po` 的翻譯時所提出（出現之處請見[此 PR](https://github.com/python/python-docs-zh-tw/pull/770#pullrequestreview-1795676679)，例如 `default cipher string`），在未討論出適當翻譯前我不希望大家的翻譯 PR 就因此被擋住（以免降低大家貢獻的意願），所以當初沒有其他疑慮就先 merge 了。

後來應該是某次我沒有參與的 sprint 有討論這個詞才將「cipher: 密碼」加入術語表（ @ken71301 @rockleona 如果還記得當初討論訂定「密碼」為適當翻譯的理由的話，再麻煩補上 🙏🏽 ）。

**Reply by SkyLull** [reply]
謝謝您的回覆 @mattwang44 


如果是您提到的 `ssl.po:255` 、 `ssl.po:259` 和 `ssl.po:259` 這三段的  `default cipher string`  的話，我認為考慮其意義後應該譯作「預設加密方法」。
### 論述
在 `ssl.po:255` 、 `ssl.po:259` 和 `ssl.po:259` 提到的 `RC4` `ChaCha20` `3DES` 都是一種加密方法，語意上應該不是將他們作為 `secret` 使用才是。

接著，在您舉例的 ssl [原始說明文件](https://docs.python.org/3/library/ssl.html#ssl.create_default_context)中提到這個 method 用於建立一個 [sslContext](https://docs.python.org/3/library/ssl.html#ssl.SSLContext)，一個用於紀錄 ssl 連線設定參數的物件。接著在 [sslContext.set_ciphers()](https://docs.python.org/3/library/ssl.html#ssl.SSLContext.set_ciphers) 裡的說明則提到欲使用這個 method 來變更 sslContext 所紀錄的加密方法，使用者的輸入應該符合 [OpenSSL cipher list format](https://docs.openssl.org/master/man1/openssl-ciphers)，而導向的 [註1] 說明頁面中的確使用了 `cipher string` 來描述在ssl工具中 `一個用於指定加密方法的字串` 
（按：`CIPHER STRINGS 
The following is a list of all permitted cipher strings and their meanings.`）

考慮到 python ssl 文內說的 `RC4` `3DES` 都是一種因不夠安全而被普遍棄用的加密方法，以及 `create_default_context()` 的作用是根據使用者指定的用途自動選擇適合的加密方法，該二句在這邊的意思應該是指這兩種方法分別在 3.4.4 和 3.6 中不再作為一種可能被自動指定於 `sslContext` 的預設 `cipher string` 選項之一。

因此我認為在這邊 `cipher` 仍作加密方法解釋。

---
註1：原始 python 文件指向的地址應該是 
> https://docs.openssl.org/master/man1/cipher/

但該頁面會自動將使用者導向

> https://docs.openssl.org/master/man1/openssl-cmds/

而此頁面的說明則表示未帶 `openssl-` 開頭的舊連結已被棄用並將遭到刪除，其功能和說明則新舊相同，故我在此貼上的連結是自行加過 `openssl-` 開頭的連結，會與實際上的 python 官方文件有出入但符合其原始意圖。

**Reply by mattwang44** [reply]
@SkyLull
感謝你詳盡的說明，是我會錯意了。我也有再看過 ssl.po 內所有其他出現 `cipher` 一詞的地方，確認都是意指「加密方法」。
也想請問你有興趣來幫忙發個 PR 修正 ssl.po 的這個錯誤嗎？

**Reply by ken71301** [reply]
@mattwang44 @SkyLull 

當天的討論應該是聚焦在 cipher 跟 secret 是不是需要語意上的分離 (主要是針對 cipher 作為加密用的 secret 時應該被翻譯成什麼 這件事)
然後我們參照劍橋 / wiki / 當天找得到的投影片資料 決定將 cipher 翻譯成 密碼 以跟 secret 分離 (當下是把兩個東西視為近似但翻譯分離)
還有參照 ciphertext(密文) 與近似詞 code(from 劍橋) 所得出的結果

我看了一下 ssl 的內文，翻成 `加密方法` 的確也沒有問題，所以問題大概就出在當時 `作為加密用的 secret 那句在哪....`
如果按照 `加密方法` 去翻的話 劍橋給的定義
a system of writing that prevents most people from understanding the message 是沒有問題的
但 如果是 wiki 的定義:
In cryptography, a cipher (or cypher) is an algorithm for performing encryption or decryption—a series of well-defined steps that can be followed as a procedure.
則其也包含解密部分的邏輯 就可能會翻成 `加解密方法`
目前想法是這樣 供參

**Reply by SkyLull** [reply]
@mattwang44 沒問題！不過我希望先完成手上的 `json.po` 再朝下一個目標進行。
